### PR TITLE
Refactor ThumbnailCtrl, create stand-alone ScrolledThumbnail widget

### PR DESCRIPTION
**Pull request by eagerm**
ThumbnailCtrl is more of a image browser demo application than a widget,
in that it reads files from a directory, selects which files to display,
deletes files, displays the source directory path in a text ctrl, etc.
This makes it unlikely that it could be used in any other application,
for example, to provide thumbnails of files with different file types
than the ones hard-coded in the class.

ThumbnailCtrl delegates most of its operations to ScrolledThumbnail
which actually implements a scrolled window of thumbnails, a Thumb class,
which contains information about a thumbnail, and an ImageHandler class,
which manipulates images.  There was poor isolation of functionality
between these classes, violating object-oriented design, with one class
making changes to the internal data of another class.  Additionally, there
was substantial non-functional code, as well as code which did not
function correctly.

This refactoring maintains the functionality and interfaces of
ThumbnailCtrl, except for those which were unused.  Existing uses of
the thumbnailctrl package should work without modification.  A new package,
scrolledthumbnail, contains the functionality for a scrolled window
containing thumbnails, an extendable Thumb class, and image manipulation
classes.  The scrolledthumbnail package can be used in other applications,
independent of the ThumbnailCtrl class, and without the functional
restrictions of that application.

Detailed changes:

ThumbnailCtrl.py (demo program):
 - Always import from wx.lib.agw
 - Optional code to use PIL instead of native image handling
 - Add setting for thumbnail width and height
 - Increase size of demo window

thumbnailctrl.py:
 - Move Thumb, ScrolledThumbnail, ImageHandler to scrolledthumbnail.py
 - Remove EVT_THUMBNAILS_CAPTION_CHANGED (unused)
 - Add EVT_THUMBNAILS_CHAR to respond to keystrokes
 - Remove image processing code
 - Add scrolling dialog for delete files
 - Move directory processing from ScrolledThumbnail
 - Move file delete processing from ScrolledThumbnail
 - List all files to be deleted in scrolling dialog
 - Remove unused or unimplemented methods and options

scrolledthumbnail.py:
 - Move Thumb, ScrolledThumbnail, ImageHander classes from thumbnailctrl.py
 - Add documentation for ScrolledThumbnail widget
 - Add example program which does not use ThumbnailCtrl
 - New EVT_THUMBNAILS_CHAR event for key press
 - Remove unused options and dead code
 - Add Rotate() to PILImageHandler and NativeImageHandler
 - Throw event EVT_THUMBNAILS_CHAR for keystroke on thumbnail
 - Fix failure to rotate images correctly
 - Redisplay window when thumb size changed
 - Simplify logic
 - Remove popup dialog when rotating images

Fixes #NNNN

### CODE REVIEW

**Comment by Metallicow** [comment]
@eagerm 
`Remove EVT_THUMBNAILS_CAPTION_CHANGED (unused)`
Don't do it. You will break software. Leave it in.

File: demo/agw/ThumbnailCtrl.py [hunk 1]
```diff
@@ -12,13 +12,13 @@
...
 
-try:
-    from agw import thumbnailctrl as TC
-except ImportError: # if it's not there locally, try the wxPython lib.
-    import wx.lib.agw.thumbnailctrl as TC
+import wx.lib.agw.thumbnailctrl as TC
 
 import images
 
+from wx.lib.agw.scrolledthumbnail import EVT_THUMBNAILS_SEL_CHANGED, EVT_THUMBNAILS_POINTED, \
```
Comment (Metallicow) [inline demo/agw/ThumbnailCtrl.py:19]: Don't use a backslash. Use parenthesis (EVT_, ...)

File: demo/agw/ThumbnailCtrl.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ def __init__(self, parent, log):
         self.statusbar = self.CreateStatusBar(2)
         self.statusbar.SetStatusWidths([-2, -1])
         # statusbar fields
-        statusbar_fields = [("ThumbnailCtrl Demo, Andrea Gavana @ 10 Dec 2005"),
+        statusbar_fields = [("ThumbnailCtrl Demo, Michael Eager @ 15 Oct 2020"),
```
Comment (Metallicow) [inline demo/agw/ThumbnailCtrl.py:35]: Dont remove someones copywrite. Just append your own to it.

**Comment by eagerm** [comment]
Thanks for the review.

On 10/20/20 12:28 PM, Metallicow wrote:
> *@Metallicow* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In demo/agw/ThumbnailCtrl.py 
> <https://github.com/wxWidgets/Phoenix/pull/1828#discussion_r508784372>:
> 
>>  
>   import images
>   
> +from wx.lib.agw.scrolledthumbnail import EVT_THUMBNAILS_SEL_CHANGED, EVT_THUMBNAILS_POINTED, \
> 
> Don't use a backslash. Use parenthesis (EVT_, ...)

OK.

-- 
Michael Eager

**Comment by eagerm** [comment]
> @eagerm
> `Remove EVT_THUMBNAILS_CAPTION_CHANGED (unused)`
> Don't do it. You will break software. Leave it in.

The original code says this is unused.  This event is never issued.  

I can leave it in, but it is just dead code.

File: demo/agw/ThumbnailCtrl.py (hunk 2 above)
Comment (eagerm) [reply demo/agw/ThumbnailCtrl.py:35]: > Dont remove someones copywrite. Just append your own to it.

The original copyright is retained.  This is the status line for the demo.

**Comment by RobinD42** [comment]
This pull request has been mentioned on **Discuss wxPython**. There might be relevant details there:

https://discuss.wxpython.org/t/refactor-thumbnailctrl-create-stand-alone-scrolledthumbnail-widget/34974/2

File: demo/agw/ThumbnailCtrl.py (hunk 1 above)
Comment (eagerm) [reply demo/agw/ThumbnailCtrl.py:19]: Done.

File: demo/agw/ThumbnailCtrl.py (hunk 2 above)
Comment (Metallicow) [reply demo/agw/ThumbnailCtrl.py:35]: @infinity77 Is this Ok with you? ... or how would you deal with this type of change...? It is status text.

Comment (infinity77) [reply demo/agw/ThumbnailCtrl.py:35]: I don’t have a problem, as long as my name is mentioned somewhere in the library 😀. I think @eagerm has done a very good job in refactoring ThumbnailCtrl into a more usable component

**Comment by Metallicow** [comment]
@eagerm As far as the EVT_ the reason to leave it in is because in a lot of Andreas lib widgets there is unused/unfinished? things. This does not mean it is unused, but Andrea may or may not have left it there so developers could hotpatch into it and assign some functionality or otherwise it might have gotten forgotten by him and should have had a TODO/NotImplemented tag on it instead of unused.
Ex: there are still a few unthemeable parts in agw aui that need getters setters for example. These will eventually get addressed when we get time to revisit the files again.

So basically in a nutshell, instead of removing it and breaking software. Do as I would do. `Break software in style!` lol
Actually implement it with some usable functionality that would be rational for developers that probably hotpatched it in their programs anyway and when it does break developers programs and they come back looking, then they will realize that they will either need to override your event if they still want their implementation or choose to adapt to your newly implemented functionality.

File: wx/lib/agw/scrolledthumbnail.py [hunk 3]
```diff
@@ -0,0 +1,2121 @@
...
+                    imgRect.height = self._tHeight
+
+            dc.SetBrush(wx.TRANSPARENT_BRUSH)
+
+            if selected:
+
+                dc.SetPen(self.grayPen)
+                dc.DrawRoundedRectangle(dotrect, 2)
+
+                dc.SetPen(wx.Pen(wx.WHITE))
```
Comment (Metallicow) [inline wx/lib/agw/scrolledthumbnail.py:1661]: wx.WHITE_PEN is faster since pen width isn't anything else but 1. It is builtin.
Same goes for BLACKPEN, GREEN_PEN, etc etc...

File: wx/lib/agw/scrolledthumbnail.py [hunk 4]
```diff
@@ -0,0 +1,2121 @@
...
+                pen.SetJoin(wx.JOIN_MITER)
+                dc.SetPen(pen)
+                if self._tOutline == THUMB_OUTLINE_FULL:
+                    dc.DrawRoundedRectangle(imgRect.x - 1, imgRect.y - 1,
+                                            imgRect.width + 3, imgRect.height + 3, 2)
+                else:
+                    dc.DrawRectangle(imgRect.x - 1, imgRect.y - 1,
+                                     imgRect.width + 3, imgRect.height + 3)
+            else:
+                dc.SetPen(wx.Pen(wx.LIGHT_GREY))
```
Comment (Metallicow) [inline wx/lib/agw/scrolledthumbnail.py:1675]: use wx.LIGHT_GREY_PEN

File: wx/lib/agw/scrolledthumbnail.py [hunk 5]
```diff
@@ -0,0 +1,2121 @@
...
+        Handles the ``wx.EVT_CHAR`` event for :class:`ThumbnailCtrl`.
+
+        :param `event`: a :class:`KeyEvent` event to be processed.
+
+        :note: You have these choices:
+
+         (1) ``d`` key rotates 90 degrees clockwise the selected thumbnails;
+         (2) ``s`` key rotates 90 degrees counter-clockwise the selected thumbnails;
+         (3) ``a`` key rotates 180 degrees the selected thumbnails;
+         (4) ``Del`` key deletes the selected thumbnails;
```
Comment (Metallicow) [inline wx/lib/agw/scrolledthumbnail.py:1987]: My eyes don't see this in the evt handler... What is this supposed to do...?
Delete the thumb from the widget, Delete the file associated with the thumb(BAD in python this is unrecoverable) or send associated file to the RecycleBin(possible better choice, recoverable)...?
Better explanation needed...

**Comment by eagerm** [comment]
On 10/20/20 10:48 PM, Metallicow wrote:
> *@Metallicow* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In wx/lib/agw/scrolledthumbnail.py 
> <https://github.com/wxWidgets/Phoenix/pull/1828#discussion_r509004775>:
> 
>> +                imgRect.x = x
> +                imgRect.y = y
> +                imgRect.width = bmp.GetWidth() - self._tBorder
> +                imgRect.height = bmp.GetHeight() - self._tBorder
> +
> +                if self._tOutline == THUMB_OUTLINE_RECT:
> +                    imgRect.height = self._tHeight
> +
> +            dc.SetBrush(wx.TRANSPARENT_BRUSH)
> +
> +            if selected:
> +
> +                dc.SetPen(self.grayPen)
> +                dc.DrawRoundedRectangle(dotrect, 2)
> +
> +                dc.SetPen(wx.Pen(wx.WHITE))
> 
> wx.WHITE_PEN is faster since pen width isn't anything else but 1. It is 
> builtin.
> Same goes for BLACKPEN, GREEN_PEN, etc etc...

Since this is a new file created from another file, the diffs show this 
as an addition, when actually it isn't.

I made no changes to the core operations of placing the tumbnail tiles 
in the window.   This is the original code.

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/20/20 10:49 PM, Metallicow wrote:
> *@Metallicow* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In wx/lib/agw/scrolledthumbnail.py 
> <https://github.com/wxWidgets/Phoenix/pull/1828#discussion_r509005098>:
> 
>> +
> +                dc.SetPen(wx.Pen(wx.WHITE))
> +                dc.DrawRectangle(imgRect.x, imgRect.y,
> +                                 imgRect.width, imgRect.height)
> +
> +                pen = wx.Pen((selected and [colour] or [wx.LIGHT_GREY])[0], 2)
> +                pen.SetJoin(wx.JOIN_MITER)
> +                dc.SetPen(pen)
> +                if self._tOutline == THUMB_OUTLINE_FULL:
> +                    dc.DrawRoundedRectangle(imgRect.x - 1, imgRect.y - 1,
> +                                            imgRect.width + 3, imgRect.height + 3, 2)
> +                else:
> +                    dc.DrawRectangle(imgRect.x - 1, imgRect.y - 1,
> +                                     imgRect.width + 3, imgRect.height + 3)
> +            else:
> +                dc.SetPen(wx.Pen(wx.LIGHT_GREY))
> 
> use wx.LIGHT_GREY_PEN

See previous comment.  This is unchanged from the original.

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/20/20 9:31 PM, Andrea Gavana wrote:
> *@infinity77* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In demo/agw/ThumbnailCtrl.py 
> <https://github.com/wxWidgets/Phoenix/pull/1828#discussion_r508982820>:
> 
>> @@ -32,7 +32,7 @@ def __init__(self, parent, log):
>           self.statusbar = self.CreateStatusBar(2)
>           self.statusbar.SetStatusWidths([-2, -1])
>           # statusbar fields
> -        statusbar_fields = [("ThumbnailCtrl Demo, Andrea Gavana @ 10 Dec 2005"),
> +        statusbar_fields = [("ThumbnailCtrl Demo, Michael Eager @ 15 Oct 2020"),
> 
> I don’t have a problem, as long as my name is mentioned somewhere in the 
> library 😀. I think @eagerm <https://github.com/eagerm> has done a very 
> good job in refactoring ThumbnailCtrl into a more usable component

Thanks, Andrea.   Your name and acknowledgement is in the source code, 
as well as the kudos you gave to others.

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/20/20 10:12 PM, Metallicow wrote:
> @eagerm <https://github.com/eagerm> As far as the EVT_ the reason to 
> leave it in is because in a lot of Andreas lib widgets there is 
> unused/unfinished? things. This does not mean it is unused, but Andrea 
> may or may not have left it there so developers could hotpatch into it 
> and assign some functionality or otherwise it might have gotten 
> forgotten by him and should have had a TODO/NotImplemented tag on it 
> instead of unused.
> Ex: there are still a few unthemeable parts in agw aui that need getters 
> setters for example. These will eventually get addressed when we get 
> time to revisit the files again.
> 
> So basically in a nutshell, instead of removing it and breaking 
> software. Do as I would do. |Break software in style!| lol
> Actually implement it with some usable functionality that would be 
> rational for developers that probably hotpatched it in their programs 
> anyway and when it does break developers programs and they come back 
> looking, then they will realize that they will either need to override 
> your event if they still want their implementation or choose to adapt to 
> your newly implemented functionality.

I retained non-functional code which was in the interface:  see the 
'thumbfilter' parameter to 'ThumbnailCtrl'.

The removed EVT_THUMBNAILS_CAPTION_CHANGED event seems unlikely to break 
any code. There is no existing code which could ever have used this 
event.  The event is never issued.  No "override the event is possible", 
since there is no event.  I can add this event to ThumbnailCtrl, as I 
mentioned, but it will be dead code.

As for adding (or retaining) unused/unfinished/broken code in a program, 
that's a problem in my opinion.  This constrains the way that the code 
can be extended or modified.  It implicitly says, this is how the code 
needs to be designed, even if the original author says explicitly, as in 
ThumbnailCtrl, he doesn't know how to make it work.  It isn't an aid to 
future extensions; it's an impediment.  (Note that this code is fifteen 
years old.  It's not as if Andrea or someone was going to "get right 
back" and complete it.)

There was a lot of unimplemented code in ThumbnailCtrl.  None of it was 
for theaming the widget.

A lot of the non-functional code in ThumbnailCtrl was intended to make a 
widget into a "more functional" application, which would make it even 
less usable in an application.

As for making unimplemented code do something usable as you suggest, 
that's not what I wanted.  I don't have any desire make ThumbnailCtrl a 
better file browser application.   I can't use it as currently designed. 
  I can use the underlying scrolled window functionality, now separated 
into ScrolledThumbnail.  My only interest with ThumbnailCtrl was to 
retain exactly the same functionality and interface as currently exists. 
  (And, incidentally, fix a number of bugs along the way.)

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/20/20 10:59 PM, Metallicow wrote:
> *@Metallicow* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In wx/lib/agw/scrolledthumbnail.py 
> <https://github.com/wxWidgets/Phoenix/pull/1828#discussion_r509008255>:
> 
>> +
> +        self.Refresh()
> +
> +
> +    def OnChar(self, event):
> +        """
> +        Handles the ``wx.EVT_CHAR`` event for :class:`ThumbnailCtrl`.
> +
> +        :param `event`: a :class:`KeyEvent` event to be processed.
> +
> +        :note: You have these choices:
> +
> +         (1) ``d`` key rotates 90 degrees clockwise the selected thumbnails;
> +         (2) ``s`` key rotates 90 degrees counter-clockwise the selected thumbnails;
> +         (3) ``a`` key rotates 180 degrees the selected thumbnails;
> +         (4) ``Del`` key deletes the selected thumbnails;
> 
> My eyes don't see this in the evt handler... What is this supposed to do...?
> Delete the thumb from the widget, Delete the file associated with the 
> thumb(BAD in python this is unrecoverable) or send associated file to 
> the RecycleBin(possible better choice, recoverable)...?
> Better explanation needed...

I'll remove the line (4) and mention that all other keystrokes cause an 
EVT_THUMBNAILS_CHAR to be thrown with the character.  T

he updated ThumbnailCtrl catches this event and deletes the file (after 
displaying an improved OK dialog box).  This functionality (bad as it 
may be) duplicates the original Thumbnailctrl.

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/21/20 8:18 AM, Michael Eager wrote:
> On 10/20/20 10:12 PM, Metallicow wrote:
>> @eagerm <https://github.com/eagerm> As far as the EVT_ the reason to 
>> leave it in is because in a lot of Andreas lib widgets there is 
>> unused/unfinished? things. This does not mean it is unused, but Andrea 
>> may or may not have left it there so developers could hotpatch into it 
>> and assign some functionality or otherwise it might have gotten 
>> forgotten by him and should have had a TODO/NotImplemented tag on it 
>> instead of unused.
>> Ex: there are still a few unthemeable parts in agw aui that need 
>> getters setters for example. These will eventually get addressed when 
>> we get time to revisit the files again.
>>
>> So basically in a nutshell, instead of removing it and breaking 
>> software. Do as I would do. |Break software in style!| lol
>> Actually implement it with some usable functionality that would be 
>> rational for developers that probably hotpatched it in their programs 
>> anyway and when it does break developers programs and they come back 
>> looking, then they will realize that they will either need to override 
>> your event if they still want their implementation or choose to adapt 
>> to your newly implemented functionality.
> 
> I retained non-functional code which was in the interface:  see the 
> 'thumbfilter' parameter to 'ThumbnailCtrl'.
> 
> The removed EVT_THUMBNAILS_CAPTION_CHANGED event seems unlikely to break 
> any code. There is no existing code which could ever have used this 
> event.  The event is never issued.  No "override the event is possible", 
> since there is no event.  I can add this event to ThumbnailCtrl, as I 
> mentioned, but it will be dead code.
> 
> As for adding (or retaining) unused/unfinished/broken code in a program, 
> that's a problem in my opinion.  This constrains the way that the code 
> can be extended or modified.  It implicitly says, this is how the code 
> needs to be designed, even if the original author says explicitly, as in 
> ThumbnailCtrl, he doesn't know how to make it work.  It isn't an aid to 
> future extensions; it's an impediment.  (Note that this code is fifteen 
> years old.  It's not as if Andrea or someone was going to "get right 
> back" and complete it.)
> 
> There was a lot of unimplemented code in ThumbnailCtrl.  None of it was 
> for theaming the widget.
> 
> A lot of the non-functional code in ThumbnailCtrl was intended to make a 
> widget into a "more functional" application, which would make it even 
> less usable in an application.
> 
> As for making unimplemented code do something usable as you suggest, 
> that's not what I wanted.  I don't have any desire make ThumbnailCtrl a 
> better file browser application.   I can't use it as currently designed. 
>   I can use the underlying scrolled window functionality, now separated 
> into ScrolledThumbnail.  My only interest with ThumbnailCtrl was to 
> retain exactly the same functionality and interface as currently exists. 
>   (And, incidentally, fix a number of bugs along the way.)

I just noticed that the unimplemented EVT_THUMBNAILS_CAPTION_CHANGED is 
still in the code.  I meant to remove it as noted in the change log, but 
it appears that I didn't.


-- 
Michael Eager

**Comment by Metallicow** [comment]
I haven't gotten around to revisiting the thumbnail ctrl, but yes refactoring out the thumbs is definitely an improvement. It is an irritation in my programs also to try and access them.
On the other hand, in my programs, I don't allow the thumbnailer to actually delete files from the hard drive. "deleting" in my case is in the sense of filtering the thumbs that is displayed in the widget(ThumbnailCtrl).
Actual deletion of files from the HDD should be implemented by the dev in my opinion.
As far as the Pen/Color stuff, that is just a hint to change that stuff to speed it up while you are mucking around with refactoring other stuff. Just some simple small changes. won't break anything. If you don't want to at the moment, that is ok. I can revisit this when I get around to doing the aui themeing stuff.

**Comment by Metallicow** [comment]
As far as the evt not being used. It is what is called hotpatching. This often happens on the develops side of a program, not in the library. Another term for that would be injection of code.

**Comment by eagerm** [comment]
On 10/21/20 8:39 AM, Metallicow wrote:
> On the other hand, in my programs, I don't allow the thumbnailer to 
> actually delete files from the hard drive. "deleting" in my case is in 
> the sense of filtering the thumbs that is displayed in the 
> widget(ThumbnailCtrl).

If you have an application using ThumbnailCtrl, can you test that it 
works with the modified version.  The only test I had is the demo.

How do you prevent ThumbnailCtrl from deleting files?  Or do you patch 
the code to fix this issue?


-- 
Michael Eager

**Comment by Metallicow** [comment]
As far as I can tell, In the old ThumbnailCtrl, this functionality was never implemented. This is why your refactoring will make it easier to allow the dev using the library to implement it themselves. Before you had to look at the source file and figure out which _protected or __private attribute to hack into to access the thumbnails, which wasn't exactly the best design... and creates more hard to maintain code on the dev's end... But your fixing that. Thanks.

If the demo shows you how to do something, then that is fine as long as any deleting of actual HDD files stays out of the library files. "delete" in my opinion in the demo should delete the thumbnail from the widget, not any actual files it is associated with on the actual HDD. By doing it this way, it shows the demo user how to create a filtering mechanism. Sometimes the user will want to filter the thumbnails before they are created and other times you will simply want to remove thumbnails from the widget as far as filtering goes.

**Comment by Metallicow** [comment]
@eagerm I tracked it down.

In the old library file this is where the deletion occurs. These two methods can be overridden as I stated.

```python
    def OnChar(self, event):
        ...
        elif event.KeyCode == wx.WXK_DELETE:
            self.DeleteFiles()

    def DeleteFiles(self):
        """
        Deletes the selected thumbnails and their associated files.

        .. warning:: This method deletes the original files too.
        """

        dlg = wx.MessageDialog(self, 'Are you sure you want to delete the files?',
                               'Confirmation',
                               wx.YES_NO | wx.NO_DEFAULT | wx.ICON_QUESTION)

        if dlg.ShowModal() == wx.ID_YES:
            errordelete = []
            count = 0

            dlg.Destroy()

            wx.BeginBusyCursor()

            for ii in range(len(self._items)):
                if self.IsSelected(ii):
                    thumb = self._items[ii]
                    files = self._items[ii].GetFullFileName()
                    filename = opj(files)
                    try:
                        os.remove(filename)
                        count = count + 1
                    except:
                        errordelete.append(files)

            wx.EndBusyCursor()

            if errordelete:
                strs = "Unable to remove the following files:\n\n"
                for fil in errordelete:
                    strs = strs + fil + "\n"
                strs = strs + "\n"
                strs = strs + "Please check your privileges and file permissions."
                dlg = wx.MessageDialog(self, strs,
                                       'Error in removing files',
                                       wx.OK | wx.ICON_ERROR)
                dlg.ShowModal()
                dlg.Destroy()

            if count:
                self.UpdateShow()
```

If you want to retain this functionality, it should probably be implemented as an option rather than default or better yet completely removed from the lib file and added to the demo. Otherwise you will have a user inadvertently press a key and then be mad at someone.
... So that Del Key might Delete/Remove the thumbnail from the widget if that option was passed to the ThumbnailCtrl.

**Comment by eagerm** [comment]
On 10/21/20 9:21 PM, Metallicow wrote:
> If the demo shows you how to do something, then that is fine as long as 
> any deleting of actual HDD files stays out of the library files. 
> "delete" in my opinion in the demo should delete the thumbnail from the 
> widget, not any actual files it is associated with on the actual HDD. By 
> doing it this way, it shows the demo user how to create a filtering 
> mechanism. Sometimes the user will want to filter the thumbnails before 
> they are created and other times you will simply want to remove 
> thumbnails from the widget as far as filtering goes.

Are you talking about demo/agw/ThumbnailCtrl.py?  The demo doesn't 
delete anything.

If you are referring to lib/agw/thumbnailctrl.py and the ThumbnailCtrl 
widget, this is existing functionality.  My intent was not to change the 
interface or behavior.  You were adamant that these remain unchanged.

As I mentioned, I'm not interested in using ThumbnailCtrl.  People (like 
me) who want to have different behavior should use ScrolledThumbnail and 
implement the behavior they desire.

-- 
Michael Eager

**Comment by eagerm** [comment]
On 10/21/20 10:01 PM, Metallicow wrote:
> @eagerm <https://github.com/eagerm> I tracked it down.
> 
> In the old library file this is where the deletion occurs. These two 
> methods can be overridden as I stated.
There was no need to track anything down.  It wasn't lost.

> If you want to retain this functionality, it should probably be 
> implemented as an option rather than default or better yet completely 
> removed from the lib file and added to the demo. Otherwise you will have 
> a user inadvertently press a key and then be mad at someone.
> ... So that Del Key might Delete/Remove the thumbnail from the widget if 
> that option was passed to the ThumbnailCtrl.

Again, I'm not making any changes in the functionality of ThumbnailCtrl.

In the existing ThumbnailCtrl, if a user hits the DEL key, they receive 
a popup which asks if they wish to delete the selected files, without 
identifying the files.  I improved this by listing the files which would 
be deleted.  That is the total extent that I intend to modify the 
existing behavior.

If users have been unhappy with the exiting behavior, they have had 15 
years to file a bug report.

You were adamant that no changes be made to the interface, now you are 
asking me to make interface and functional changes to a control which I 
have no interest in using or maintaining.  Certainly you are welcome to 
make whatever design changes you wish and to submit a pull request.

-- 
Michael Eager

**Comment by Metallicow** [comment]
I hotpatch/inject in my programs to avoid the os.remove, so keeping that in for this PullREQz is fine since all you aim to do is refactor.
But yes, at some point when I get to revisiting/improving it, I will pull my code out of my program and send a PullREQz to change that behaviour to just rationally remove the thumbnail from the widget, since that will make it easier so you don't have to dirty your refactor up too much.
At which point the new functionality will just replace the bad design choice and stick the DeleteFile def into the demo where it should have been in the first place and show users how to override it in their programs if they really want to retain that. Of course as I said also, the recyclebin is a recoverable delete as where os.remove is not.

Anyhow, If you are ready, then I really don't see a whole lot more that would need to be done before merging, unless you have other questions... Just make sure the library files and the wxPy demo file works without errors and you should be good to go.
You may want to add another demo to the wxPy demo framework to show that you can now use thumbnails separate and also add that to the `recent changes` Tree.

**Comment by eagerm** [comment]
> Anyhow, If you are ready, then I really don't see a whole lot more that would need to be done before merging, unless you have other questions... Just make sure the library files and the wxPy demo file works without errors and you should be good to go.

Everything works as far as I can tell.  Some things better than before.  :-)

> You may want to add another demo to the wxPy demo framework to show that you can now use thumbnails separate and also add that to the `recent changes` Tree.

I'll think about another demo.  Maybe duplicate most of the current demo, bypassing ThumbnailCtrl.  

What and where is the recent changes Tree?

**Comment by Metallicow** [comment]
@eagerm 
To add a new demo, add probably just a simple demo in the agw directory called Thumbnail.py and put just a thumbnail or 2 on a frame/panel etc...
Then to add it to the framework tree add the filename here.
https://github.com/wxWidgets/Phoenix/blob/master/demo/demodata.py

**Comment by eagerm** [comment]
On 10/22/20 3:35 PM, Metallicow wrote:
> @eagerm <https://github.com/eagerm>
> To add a new demo, add probably just a simple demo in the agw directory 
> called Thumbnail.py and put just a thumbnail or 2 on a frame/panel etc...
> Then to add it to the framework tree add the filename here.
> https://github.com/wxWidgets/Phoenix/blob/master/demo/demodata.py

I think you mean demo/agw/__demo__.py, where ThumbnailCtrl is listed.

I'll create a minimal ScrolledThumbnail.py demo.

-- 
Michael Eager

**Comment by Metallicow** [comment]
Well, yes, since it is a new addition/change then there would be two entries for the demo data.

Also Robin will ask you to make a change.log entry.
`* Refactored Thumbnail out of agw.ThumbnailCtrl so as to be usable outside of ThumbnailCtrl. #1828`
should be fine for a change log line.

**Comment by eagerm** [comment]
On 10/22/20 4:41 PM, Metallicow wrote:
> Well, yes, since it is a new addition/change then there would be two 
> entries for the demo data.

I'm not following you.  What two entries?

-- 
Michael Eager

**Comment by Metallicow** [comment]
This is where the 1st one should go, and the other would reside in the agw tree section of demo data.
The Recent additions/updates is first to notify users of the changes since Phoenix/4.0, etc and will reset when 4.2/4.3 is released.
```
_treeList = [
    # new stuff
    ('Recent Additions/Updates', [
        'Simplebook',
        'LEDNumberCtrl',
        'DynamicSashWindow',
        'TreeListCtrl',
        'NotificationMessage',
        'AddPrivateFont',
        'SVGImage_Bitmap',
        'SVGImage_Render',
        'ActivityIndicator',
        'GenericCheckBox',
        'CheckListCtrl',
    ]),
```

**Comment by Metallicow** [comment]
If you have issues with the recent additions tree since agw is a package, then just don't worry about it.
When I get the new framework done it will cure the multiple package issue with users packages not showing.
The current implementation only allows for agw event ho it says otherwise. It was a bit of a hiccup when he designed the framework years ago.

**Comment by eagerm** [comment]
I've created a new demo:  demo/agw/ScrolledThumbnail.py.   It has almost the same functionality as the ThumbnailCtrl demo and uses much of the same code.  I also added it to the Recent and AGW lists.   Also updated the CHANGES file.

**Comment by eagerm** [comment]
Hi --

I've resolved all of the comments on the pull request.  I also created a 
demo program for ScrolledThumbnail widget.

I noticed that GitHub flags the added demo program with the comment 
"This commit does not belong to any branch on this repository."  I'm not 
sure what this means.  The file has been added to the "master" branch in 
my GitHub repo.  Is this because it does not in the wxWidgets repo?

What are the next steps?

On 10/20/20 2:39 PM, Robin Dunn via Discuss wxPython wrote:
> [Robin] 	Robin <https://discuss.wxpython.org/u/robin>
> October 20
> 
> For the record, the PR is here:
> 
> github.com/wxWidgets/Phoenix 
> <https://github.com/wxWidgets/Phoenix/pull/1828>
> 
> 
>         Refactor ThumbnailCtrl, create stand-alone ScrolledThumbnail
>         widget <https://github.com/wxWidgets/Phoenix/pull/1828>
> 
> |wxWidgets:master| ← |eagerm:master|
> opened 06:16PM - 20 Oct 20 UTC
> eagerm eagerm <https://github.com/eagerm>
> +2264 -2183 <https://github.com/wxWidgets/Phoenix/pull/1828/files>
> ------------------------------------------------------------------------
> 
> Visit Topic 
> <https://discuss.wxpython.org/t/refactor-thumbnailctrl-create-stand-alone-scrolledthumbnail-widget/34974/2> 
> or reply to this email to respond.
> 
> To unsubscribe from these emails, click here 
> <https://discuss.wxpython.org/email/unsubscribe/7e0e8832f8a60ee9398420a4d13c6fdd9f15fd91987b45bc3aea7ed91bc2815c>.
> 

-- 
Michael Eager

**Comment by RobinD42** [comment]
There is something very wrong with this PR, it is undoing a lot of the recent changes throughout the source tree. Please redo this PR like the following. If you need help with it search for "github workflow" for more information:

1. In your local repository make a **new branch**  starting from the **current master** in this repository (your upstream.) Currently that is revision 8a26e17b168f831a846772e9400e2cea6a3b291d.

2. Redo your changes in that local branch

3. Push your branch to your origin remote (aka your fork on github)

4. Create your PR from that branch. When you do, add a link back to this PR so the history and discussion can be maintained.

**Comment by eagerm** [comment]
On 10/28/20 9:14 AM, Robin Dunn wrote:
> There is something very wrong with this PR, it is undoing a lot of the 
> recent changes throughout the source tree.

When I updated the Changelog, it had a merge conflict.  To resolve this 
(which I don't think worked) I did a pull from the upstream repo and 
merged it in to my repo.  That probably was a mistake.

  Please redo this PR like the
> following. If you need help with it search for "github workflow" for 
> more information:
> 
>  1. In your local repository make a *new branch* starting from the
>     *current master* in this repository (your upstream.) Currently that
>     is revision 8a26e17
>     <https://github.com/wxWidgets/Phoenix/commit/8a26e17b168f831a846772e9400e2cea6a3b291d>.
> 
>  2. Redo your changes in that local branch
> 
>  3. Push your branch to your origin remote (aka your fork on github)
> 
>  4. Create your PR from that branch. When you do, add a link back to
>     this PR so the history and discussion can be maintained.

I'll do that.

-- 
Michael Eager

**Comment by eagerm** [comment]
Created new pull request https://github.com/wxWidgets/Phoenix/pull/1836

On 10/28/20 9:27 AM, Michael Eager wrote:
> On 10/28/20 9:14 AM, Robin Dunn wrote:
>> There is something very wrong with this PR, it is undoing a lot of the 
>> recent changes throughout the source tree.
> 
> When I updated the Changelog, it had a merge conflict.  To resolve this 
> (which I don't think worked) I did a pull from the upstream repo and 
> merged it in to my repo.  That probably was a mistake.
> 
>   Please redo this PR like the
>> following. If you need help with it search for "github workflow" for 
>> more information:
>>
>>  1. In your local repository make a *new branch* starting from the
>>     *current master* in this repository (your upstream.) Currently that
>>     is revision 8a26e17
>>     
>> <https://github.com/wxWidgets/Phoenix/commit/8a26e17b168f831a846772e9400e2cea6a3b291d>. 
>>
>>
>>  2. Redo your changes in that local branch
>>
>>  3. Push your branch to your origin remote (aka your fork on github)
>>
>>  4. Create your PR from that branch. When you do, add a link back to
>>     this PR so the history and discussion can be maintained.
> 
> I'll do that.
> 

-- 
Michael Eager
